// scroll-top/esm2020/lib/scroll-top.component.mjs
import { Component, Input } from '@angular/core';
import { ScrollTopDirective } from './scroll-top.directive';
import * as i0 from "@angular/core";
import * as i1 from "./scroll-top.service";
export class ScrollTopComponent {
    constructor(service) {
        this.service = service;
    }
    set elementId(value) {
        this._elementId = value;
        this.service.init(value);
    }
    ngOnInit() {
        this.service.init(this._elementId);
    }
    ngOnDestroy() {
        this.service.destroy();
    }
}
ScrollTopComponent.ɵfac = function ScrollTopComponent_Factory(t) { return new (t || ScrollTopComponent)(i0.ɵɵdirectiveInject(i1.ScrollTopService)); };
ScrollTopComponent.ɵcmp = /*@__PURE__*/ i0.ɵɵdefineComponent({ type: ScrollTopComponent, selectors: [["lib-scroll-top"]], inputs: { elementId: "elementId" }, features: [i0.ɵɵProvidersFeature([ScrollTopDirective])], decls: 1, vars: 0, consts: [[1, "scroll-top"]], template: function ScrollTopComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelement(0, "div", 0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ScrollTopComponent, [{
        type: Component,
        args: [{
                selector: 'lib-scroll-top',
                template: `
    <div class="scroll-top" ></div>
  `,
                providers: [ScrollTopDirective]
            }]
    }], function () { return [{ type: i1.ScrollTopService }]; }, { elementId: [{
            type: Input
        }] }); })();
