// scroll-top/esm2020/lib/scroll-top.service.mjs
import { Injectable, isDevMode } from '@angular/core';
import * as i0 from "@angular/core";
export class ScrollTopService {
    constructor() {
        this.isDevMode = isDevMode();
        this.isWindow = typeof window !== 'undefined';
        this._init = false;
        this.lastY = 0;
        this.ticking = false;
        this.log = (s, ...optional) => s && this.isDevMode && console.log(s, optional);
        this.handleScroll = () => {
            if (!this.ticking) {
                window.requestAnimationFrame(() => {
                    this.updateLastY();
                    this.ticking = false;
                });
                this.ticking = true;
            }
        };
    }
    get currentPositionY() {
        return this._elementToScroll ? this._elementToScroll.scrollTop : window.pageYOffset;
    }
    get info() { return Math.ceil(this.lastY); }
    setDevMode() { this.isDevMode = true; }
    init(elementId) {
        if (this.isWindow) {
            this.destroy();
            this._elementToScroll = null;
            if (elementId) {
                this._elementToScroll = document.querySelector(`#${elementId}`);
            }
            this._host = this._elementToScroll || window;
            this._host.addEventListener('scroll', this.handleScroll);
            this.log(`'NgScrolltopService' to use ${this._elementToScroll ? 'element' : 'window'}`, elementId || '');
        }
        this._init = true;
    }
    destroy() {
        this.isWindow && this._host && this._host.removeEventListener('scroll', this.handleScroll);
    }
    scrollABit(newY) {
        if (this._elementToScroll) {
            this._elementToScroll.scrollTop = newY;
        }
        else {
            window.scroll(0, newY);
        }
    }
    updateLastY() {
        if (!this._init)
            return;
        const newY = this.currentPositionY;
        if (newY !== this.lastY) {
            this.lastY = newY;
        }
    }
}
ScrollTopService.ɵfac = function ScrollTopService_Factory(t) { return new (t || ScrollTopService)(); };
ScrollTopService.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: ScrollTopService, factory: ScrollTopService.ɵfac, providedIn: 'root' });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ScrollTopService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
