// scroll-top/esm2020/lib/scroll-top.directive.mjs
import { Directive, HostListener, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./scroll-top.service";
const DURATION = 1000;
export class ScrollTopDirective {
    constructor(service, ngZone) {
        this.service = service;
        this.ngZone = ngZone;
        this.progress = 0;
        this.easing = (x) => {
            'use strict';
            if (x < 0.5) {
                return Math.pow(x * 2, 2) / 2;
            }
            return 1 - Math.pow((1 - x) * 2, 2) / 2;
        };
        this.scrollABit = this.scrollABit.bind(this);
    }
    onclick($event) {
        this.startTop = this.service.currentPositionY;
        this.startTime = null;
        if (this.targetElementExists($event, this.targetElement)) {
            this.ngZone.runOutsideAngular(() => { window.requestAnimationFrame(this.scrollABit); });
        }
        this.service.log('NgScrolltopDirective click');
    }
    targetElementExists($event, selector) {
        let found = false;
        if ($event?.target.matches(`#${selector}, #${selector} > *`)) {
            found = true;
        }
        return found;
    }
    scrollABit(timestamp) {
        if (!this.startTime) {
            this.startTime = timestamp;
        }
        this.progress = timestamp - this.startTime;
        const percent = (this.progress >= DURATION ? 1 : this.easing(this.progress / DURATION));
        const newY = this.startTop - Math.ceil(this.startTop * percent);
        this.service.scrollABit(newY);
        if (percent < 1) {
            this.ngZone.runOutsideAngular(() => { window.requestAnimationFrame(this.scrollABit); });
        }
    }
}
ScrollTopDirective.ɵfac = function ScrollTopDirective_Factory(t) { return new (t || ScrollTopDirective)(i0.ɵɵdirectiveInject(i1.ScrollTopService), i0.ɵɵdirectiveInject(i0.NgZone)); };
ScrollTopDirective.ɵdir = /*@__PURE__*/ i0.ɵɵdefineDirective({ type: ScrollTopDirective, selectors: [["", "blScrolltop", ""]], hostBindings: function ScrollTopDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("click", function ScrollTopDirective_click_HostBindingHandler($event) { return ctx.onclick($event); });
    } }, inputs: { targetElement: "targetElement" } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ScrollTopDirective, [{
        type: Directive,
        args: [{
                selector: '[blScrolltop]'
            }]
    }], function () { return [{ type: i1.ScrollTopService }, { type: i0.NgZone }]; }, { targetElement: [{
            type: Input,
            args: ['targetElement']
        }], onclick: [{
            type: HostListener,
            args: ['click', ['$event']]
        }] }); })();
