// scroll-top/esm2020/lib/scroll-top.component.mjs
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./scroll-top.service";
import * as i2 from "./scroll-top.directive";
export class ScrollTopComponent {
    constructor(service) {
        this.service = service;
    }
    set elementId(value) {
        this._elementId = value;
        this.service.init(value);
    }
    ngOnInit() {
        this.service.init(this._elementId);
    }
    ngOnDestroy() {
        this.service.destroy();
    }
}
ScrollTopComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopComponent, deps: [{ token: i1.ScrollTopService }], target: i0.ɵɵFactoryTarget.Component });
ScrollTopComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.2.0", type: ScrollTopComponent, selector: "lib-scroll-top", inputs: { elementId: "elementId" }, ngImport: i0, template: `
    <div class="scroll-top" blScrolltop>
  </div>
  `, isInline: true, dependencies: [{ kind: "directive", type: i2.ScrollTopDirective, selector: "[blScrolltop]" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'lib-scroll-top',
                    template: `
    <div class="scroll-top" blScrolltop>
  </div>
  `,
                }]
        }], ctorParameters: function () { return [{ type: i1.ScrollTopService }]; }, propDecorators: { elementId: [{
                type: Input
            }] } });
