// scroll-top/esm2020/lib/scroll-top.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ScrollTopComponent } from './scroll-top.component';
import { ScrollTopDirective } from './scroll-top.directive';
import * as i0 from "@angular/core";
export class ScrollTopModule {
}
ScrollTopModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ScrollTopModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopModule, declarations: [ScrollTopComponent,
        ScrollTopDirective], imports: [CommonModule], exports: [ScrollTopComponent,
        ScrollTopDirective] });
ScrollTopModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ScrollTopComponent,
                        ScrollTopDirective
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        ScrollTopComponent,
                        ScrollTopDirective
                    ]
                }]
        }] });
