// scroll-top/esm2020/lib/scroll-top.directive.mjs
import { Directive, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./scroll-top.service";
const DURATION = 1000;
export class ScrollTopDirective {
    constructor(service, ngZone) {
        this.service = service;
        this.ngZone = ngZone;
        this.progress = 0;
        this.easing = (x) => {
            'use strict';
            if (x < 0.5) {
                return Math.pow(x * 2, 2) / 2;
            }
            return 1 - Math.pow((1 - x) * 2, 2) / 2;
        };
        this.scrollABit = this.scrollABit.bind(this);
    }
    onclick() {
        this.startTop = this.service.currentPositionY;
        this.startTime = null;
        this.ngZone.runOutsideAngular(() => { window.requestAnimationFrame(this.scrollABit); });
        this.service.log('NgScrolltopDirective click');
    }
    scrollABit(timestamp) {
        if (!this.startTime) {
            this.startTime = timestamp;
        }
        this.progress = timestamp - this.startTime;
        const percent = (this.progress >= DURATION ? 1 : this.easing(this.progress / DURATION));
        const newY = this.startTop - Math.ceil(this.startTop * percent);
        this.service.scrollABit(newY);
        if (percent < 1) {
            this.ngZone.runOutsideAngular(() => { window.requestAnimationFrame(this.scrollABit); });
        }
    }
}
ScrollTopDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopDirective, deps: [{ token: i1.ScrollTopService }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
ScrollTopDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.2.0", type: ScrollTopDirective, selector: "[blScrolltop]", host: { listeners: { "click": "onclick()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.0", ngImport: i0, type: ScrollTopDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[blScrolltop]'
                }]
        }], ctorParameters: function () { return [{ type: i1.ScrollTopService }, { type: i0.NgZone }]; }, propDecorators: { onclick: [{
                type: HostListener,
                args: ['click']
            }] } });
