// scroll-top/esm2020/lib/scroll-top.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ScrollTopComponent } from './scroll-top.component';
import { ScrollTopDirective } from './scroll-top.directive';
import * as i0 from "@angular/core";
export class ScrollTopModule {
}
ScrollTopModule.ɵfac = function ScrollTopModule_Factory(t) { return new (t || ScrollTopModule)(); };
ScrollTopModule.ɵmod = /*@__PURE__*/ i0.ɵɵdefineNgModule({ type: ScrollTopModule });
ScrollTopModule.ɵinj = /*@__PURE__*/ i0.ɵɵdefineInjector({ imports: [CommonModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(ScrollTopModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ScrollTopComponent,
                    ScrollTopDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ScrollTopComponent,
                    ScrollTopDirective
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(ScrollTopModule, { declarations: [ScrollTopComponent,
        ScrollTopDirective], imports: [CommonModule], exports: [ScrollTopComponent,
        ScrollTopDirective] }); })();
